File: LeetCode - Algorithm/Linked List/138.java
// 138. Copy List with Random Pointer
// https://leetcode.com/problems/copy-list-with-random-pointer/

import java.util.Map;
import java.util.HashMap;

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    Map<Node, Node> hashMap;

    public Node copyRandomList(Node head) {
        if (head == null)
            return head;

        this.hashMap = new HashMap<>();

        return this.copy(head);
    }

    private Node copy(Node node) {
        if (hashMap.containsKey(node))
            return hashMap.get(node);

        Node newNode = new Node(node.val);
        hashMap.put(node, newNode);

        if (node.next != null)
            newNode.next = this.copy(node.next);
        if (node.random != null)
            newNode.random = this.copy(node.random);

        return newNode;
    }
}

File: LeetCode - Algorithm/Linked List/138.js
// 138. Copy List with Random Pointer
// https://leetcode.com/problems/copy-list-with-random-pointer/

function _Node(val, next, random) {
  this.val = val;
  this.next = next;
  this.random = random;
}

/**
 * @param {_Node} head
 * @return {_Node}
 */
var copyRandomList = function (head) {
  const hash = new Map();

  const copyNode = (node) => {
    if (!node) return null;
    if (hash.has(node)) return hash.get(node);

    const newNode = new _Node(node.val);
    hash.set(node, newNode);
    newNode.next = copyNode(node.next);
    newNode.random = copyNode(node.random);

    return newNode;
  };

  return copyNode(head);
};
